sale_purchase_force_vendor/views/sale_order_view.xml

```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="view_order_form" model="ir.ui.view">
+        <field name="name">sale.order.form</field>
+        <field name="model">sale.order</field>
+        <field name="inherit_id" ref="sale.view_order_form" />
+        <field name="arch" type="xml">
+            <xpath
+                expr="//field[@name='order_line']/tree/field[@name='price_unit']"
+                position="before"
+            >
+                <field name="allowed_vendor_ids" invisible="1" />
+                <field name="sale_purchase_force_vendor_restrict" invisible="1" />
+                <field
+                    name="vendor_id"
+                    attrs="{'invisible': ['|', ('route_id', '=', False), ('sale_purchase_force_vendor_restrict', '=', False)]}"
```

Comment:
The question is not to totally hide the field, but restrict the list of possible elements to select or not. You can add 2 times the same field, one with domain and the other without, and put visibility depending on the parameter.

Another incorrect thing is to only check the visibility by route. Not all routes require vendor.

Reply:
Ok done, the field is added 2 times and it is shown with the domain according to configuration.